File: Grunt/Grunt/Models/Waypoint/Article.cs
// <copyright file="Article.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Grunt is managed by 343 Industries and Microsoft. This wrapper is not endorsed by 343 Industries or Microsoft.
// </copyright>

using System;
using System.Collections.Generic;
using static System.Net.WebRequestMethods;

namespace OpenSpartan.Grunt.Models.Waypoint
{
    /// <summary>
    /// Article published on <see href="https://www.halowaypoint.com/">Halo Waypoint</see>.
    /// </summary>
    [IsAutomaticallySerializable]
    public class Article
    {
        /// <summary>
        /// Gets or sets the article ID.
        /// </summary>
        public int Id { get; set; }

        /// <summary>
        /// Gets or sets the URI to the featured image.
        /// </summary>
        public string? FeaturedImageUri { get; set; }

        /// <summary>
        /// Gets or sets the featured image alternative text.
        /// </summary>
        public string? FeaturedImageAlt { get; set; }

        /// <summary>
        /// Gets or sets the article title.
        /// </summary>
        public string? Title { get; set; }

        /// <summary>
        /// Gets or sets the article subtitle.
        /// </summary>
        public string? Subtitle { get; set; }

        /// <summary>
        /// Gets or sets the article content.
        /// </summary>
        public string? Content { get; set; }

        /// <summary>
        /// Gets or sets the article excerpt.
        /// </summary>
        public string? Excerpt { get; set; }

        /// <summary>
        /// Gets or sets the article slug.
        /// </summary>
        public string? Slug { get; set; }

        /// <summary>
        /// Gets or sets the article creator slug.
        /// </summary>
        public string? CreatorSlug { get; set; }

        /// <summary>
        /// Gets or sets the article creator title.
        /// </summary>
        public string? CreatorTitle { get; set; }

        /// <summary>
        /// Gets or sets the categories associated with the article.
        /// </summary>
        public List<int>? Categories { get; set; }

        /// <summary>
        /// Gets or sets the tags associated with the article.
        /// </summary>
        public List<string>? Tags { get; set; }

        /// <summary>
        /// Gets or sets article blocks that represent individual paragraphs.
        /// </summary>
        public List<ArticleBlock>? Blocks { get; set; }

        /// <summary>
        /// Gets or sets the article publish date and time.
        /// </summary>
        public DateTime? PublishDate { get; set; }

        /// <summary>
        /// Gets or sets the medium-sized featured image.
        /// </summary>
        public string? FeaturedImageUriMedium { get; set; }

        /// <summary>
        /// Gets or sets the medium-sized featured image alternative text.
        /// </summary>
        public string? FeaturedImageAltMedium { get; set; }

        /// <summary>
        /// Gets or sets the msmall-sized featured image.
        /// </summary>
        public string? FeaturedImageUriSmall { get; set; }

        /// <summary>
        /// Gets or sets the small-sized featured image alternative text.
        /// </summary>
        public string? FeaturedImageAltSmall { get; set; }
    }
}

File: src/node/src/models/waypoint/article.ts
/**
 * Article block attribute.
 */
export interface ArticleAttribute {
  /** Attribute identifier */
  id?: number;
  /** Size slug */
  sizeSlug?: string;
  /** Link destination */
  linkDestination?: string;
  /** URL */
  url?: string;
  /** Type */
  type?: string;
  /** Provider name slug */
  providerNameSlug?: string;
  /** Class name */
  className?: string;
}

/**
 * Article content block.
 */
export interface ArticleBlock {
  /** Block name */
  blockName?: string;
  /** Block attributes */
  attrs?: ArticleAttribute[];
  /** Inner blocks */
  innerBlocks?: string[];
  /** Inner HTML content */
  innerHTML?: string;
  /** Inner content */
  innerContent?: string[];
}

/**
 * Waypoint article.
 */
export interface Article {
  /** Article identifier */
  id?: number;
  /** Featured image URL */
  featuredImageUri?: string;
  /** Featured image alt text */
  featuredImageAlt?: string;
  /** Article title */
  title?: string;
  /** Article subtitle */
  subtitle?: string;
  /** Full content (HTML) */
  content?: string;
  /** Short excerpt */
  excerpt?: string;
  /** URL slug */
  slug?: string;
  /** Creator slug */
  creatorSlug?: string;
  /** Creator title */
  creatorTitle?: string;
  /** Category IDs */
  categories?: number[];
  /** Tags */
  tags?: string[];
  /** Content blocks */
  blocks?: ArticleBlock[];
  /** Publish date (ISO 8601) */
  publishDate?: string;
  /** Medium featured image URL */
  featuredImageUriMedium?: string;
  /** Medium featured image alt */
  featuredImageAltMedium?: string;
  /** Small featured image URL */
  featuredImageUriSmall?: string;
  /** Small featured image alt */
  featuredImageAltSmall?: string;
}

/**
 * Article category.
 */
export interface ArticleCategory {
  /** Category identifier */
  id?: number;
  /** Category name */
  name?: string;
  /** Category description */
  description?: string;
  /** URL slug */
  slug?: string;
  /** Number of articles */
  count?: number;
  /** Parent category ID (0 for top-level) */
  parent?: number;
}
